File: crocodile-streams/src/main/scala/fr/xebia/ldi/crocodile/stream/operator/Mappers.scala
package fr.xebia.ldi.crocodile.stream.operator

import fr.xebia.ldi.crocodile.common.model.Client.ClientKey
import fr.xebia.ldi.crocodile.common.model._
import fr.xebia.ldi.crocodile.stream.model
import fr.xebia.ldi.crocodile.stream.model.Pageview
import io.circe.parser.parse
import org.apache.kafka.common.utils.Bytes
import org.apache.kafka.streams.KeyValue
import org.apache.kafka.streams.kstream.{KeyValueMapper, ValueMapper}

import scala.util.Try

/**
  * Created by loicmdivad.
  */
object Mappers {

  val getPrice: ValueMapper[String, Int] = (value: String) =>

    parse(value).toOption.flatMap {
      _.hcursor
        .downField("price")
        .as[Double]
        .toOption
        .map(_ * 100 toInt)
    }.getOrElse(-1)

  val extractClient: ValueMapper[Link, Client] = (valueLink: Link) => valueLink.client

  val extractItems: ValueMapper[Purchase, Iterable[Item]] = (purchase: Purchase) => purchase.items

  val parseHeaders: KeyValueMapper[Bytes, Click, KeyValue[ClientKey, Pageview]] = (_, click) => (for {

      code <- click.codeOption

      login <- click.loginOption

      device <- click.deviceOption

      datetime <- click.visitTimeOption

      _ <- Try(assert(code.length equals 12)).toOption // and any other business logic rules you'd like

    } yield KeyValue.pair(ClientKey(login), model.Pageview(code, device, datetime))) orNull
}

File: kstreams-test-java/crocodile4j-streams/src/test/java/fr/xebia/ldi/crocodilej/stream/operator/Mappers.java
package fr.xebia.ldi.crocodilej.stream.operator;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import org.apache.kafka.streams.kstream.ValueMapper;

/**
 * Created by loicmdivad.
 */
public class Mappers {

    public static final ValueMapper<String, Integer> getPrice = (String json) ->
            new Gson().fromJson(json, JsonElement.class)
                    .getAsJsonObject()
                    .get("price")
                    .getAsInt() * 100;
}
